src/api/cluster_create.go

```diff
@@ -236,22 +261,39 @@ func (a *ClusterAPI) createCluster(
 		commonCluster, err = a.clusterManager.CreateCluster(ctx, creationCtx, creator)
 	}
 
-	if err == cluster.ErrAlreadyExists || isInvalid(err) {
-		logger.Debugf("invalid cluster creation: %s", err.Error())
-
-		return nil, &pkgCommon.ErrorResponse{
-			Code:    http.StatusBadRequest,
+	if err != nil {
+		internalServerErrorResponse := &pkgCommon.ErrorResponse{
+			Code:    http.StatusInternalServerError,
 			Message: err.Error(),
 			Error:   err.Error(),
 		}
-	} else if err != nil {
-		logger.Errorf("error during cluster creation: %s", err.Error())
 
-		return nil, &pkgCommon.ErrorResponse{
-			Code:    http.StatusInternalServerError,
+		badRequestErrorResponse := &pkgCommon.ErrorResponse{
+			Code:    http.StatusBadRequest,
 			Message: err.Error(),
 			Error:   err.Error(),
 		}
+
+		if err == cluster.ErrAlreadyExists {
+			logger.Debugf("invalid cluster creation: %s", err.Error())
+
+			existingCluster, getErr := a.clusterManager.GetClusterByName(ctx, organizationID, createClusterRequest.Name)
+			if getErr != nil {
+				return nil, internalServerErrorResponse
+			}
+
+			return existingCluster, badRequestErrorResponse
+		}
+
+		if isInvalid(err) {
+			logger.Debugf("invalid cluster creation: %s", err.Error())
+
+			return nil, badRequestErrorResponse
+		}
+
+		logger.Errorf("error during cluster creation: %s", err.Error())
+
+		return nil, internalServerErrorResponse
```

Comment:
**Recommendation**: there are a couple of things IMO we **SHOULD** do differently here. I'm going to provide a code suggestion at the end (but not as a GH suggestion, cause that doesn't work on deleted lines).

First I would advise never returning both a value and an error as on line 285 `return existingCluster, badRequestErrorResponse`, because it is very un-idiomatic in Go and would surprise a lot of Go engineers. When there is an error, the cluster shouldn't be created and thus returned, and where it is returned, the error should be nil.
Instead the raw already exists response should be returned and the cluster should be queried by the caller in case of an already existing error for the additional info and deciding the response.

Secondly, I would say the logical branching is a little convoluted here, with multiple branches ending up in similar result (`return nil, internalServerErrorResponse`, `return x, badRequestErrorResponse`), those could be straightened out and simplified in my opinion, I'm going to give it a try.

While I was trying to write this code I was struck by the issue of trying to return an error response from this function but also handling a non-error case here. IMO the best approach would be to query the cluster by name before calling this function and handling create on creating case before arriving here, that way this part can be simplified into handling the already exists case as having non-compatible status and thus returning conflict instead. 
What do you think? 
(an alternative I considered is adding an interface which is either a success response or an error response optionally or using an error response to encode a successful status code response, because 1 level upward the c.JSON() call seems to be very similar for non-error and error cases, but this short-cutting idea on the `CreateCluster()` function seems to be the easiest/best approach and it also sounds like the most most performant one as it doesn't rely try and fail logic for short-cutting the already existing and creating case).

Reply:
If we check whether the cluster exists or not even before calling `clusterCreate`, wouldn't it make sense to handle all "cluster already exists" cases there, and delete all code that checks for cluster existence within `clusterCreate`?

Reply:
It most definitely would, good idea.
It is worth checking the cluster existence beforehand then and if it exists with non-creating status then return the already exists error, if it exists with creating status, return the accepted success and if it doesn't exist then execute the cluster creation.

Reply:
I went with checking the cluster existence before calling cluster create as we discussed, also removed the existence checks after this point 👍